include/libtorrent/extensions.hpp

```diff
@@ -437,8 +437,20 @@ namespace libtorrent {
 		virtual bool on_reject(peer_request const&) { return false; }
 		virtual bool on_suggest(piece_index_t) { return false; }
 
+		virtual void sent_have_all() {}
+		virtual void sent_have_none() {}
+		virtual void sent_reject_request(piece_index_t, int, int) {}
+		virtual void sent_allow_fast(piece_index_t) {}
+		virtual void sent_suggest(piece_index_t) {}
+		virtual void sent_cancel(piece_index_t, int, int) {}
+		virtual void sent_request(piece_index_t, int, int) {}
+		virtual void sent_choke() {}
 		// called after a choke message has been sent to the peer
 		virtual void sent_unchoke() {}
+		virtual void sent_interested() {}
+		virtual void sent_not_interested() {}
+		virtual void sent_have(int) {}
+		virtual void sent_piece(piece_index_t, int, int) {}
```

Comment:
a concern I have with this is that there will be a lot of unnecessary virtual function calls to plugins that don't care, and it will scale poorly if most plugins are uninterested in most of these hooks. The approach I've taken with some other hooks on the (session-level) `plugin` is to have plugins "subscribe" to certain events via the `implemented_features()` bitmask. The session keeps separate lists of *just* the plugins that care about certain hooks. Perhaps this would be overkill at the peer-level plugins. I believe both the `ut_metadata` and `smart_ban` plugins installs peer connection plugins though, so there would be some more looping and making indirect function calls just by introducing these. Is there a simpler way to mitigate this concern?

Reply:
Is it significantly costly to call virtual functions which haven't been overridden?

If there's a way to check if the virtual function is overridden at subscription time, that might be easier than requiring the `implemented_features()` bitmask. (something like `&peer_plugin::sent_choke != &foo::sent_choke`)

Reply:
Unfortunately C++ does not provide any way to compare individual pointers in an object's vtable. You can only test for the presence of an entire class interface with `dynamic_cast`.

Reply:
Per [Stackoverflow](https://stackoverflow.com/a/37588766):

```c++
template <class derived_peer_plugin>
struct peer_plugin {
    virtual void on_event() {}

    peer_plugin() {
        if (&derived_peer_plugin::on_event == &peer_plugin::on_event) {
            std::cout << "not overridden" << std::endl;
        } else {
            std::cout << "overridden" << std::endl;
        }
    }
};

struct a_peer_plugin : peer_plugin<a_peer_plugin> {
    virtual void on_event() {}
};

struct b_peer_plugin : peer_plugin<b_peer_plugin> {
    // no override
};

int main()
{
    a_peer_plugin a;
    b_peer_plugin b;
}
```

=>

```
overridden
not overridden
```

Reply:
this will not fly, what do you think is the relationship between `a_peer_plugin` and `b_peer_plugin` in terms of pure OOP?

Reply:
:P

```c++
struct base_peer_plugin {
public:
    virtual ~base_peer_plugin() {}
};

template <class derived_peer_plugin>
struct peer_plugin : public base_peer_plugin {
    virtual void on_event() {}

    peer_plugin() {
        if (&derived_peer_plugin::on_event == &peer_plugin::on_event) {
            std::cout << "not overridden" << std::endl;
        } else {
            std::cout << "overridden" << std::endl;
        }
    }
};

struct a_peer_plugin : peer_plugin<a_peer_plugin> {
    virtual void on_event() {}
};

struct b_peer_plugin : peer_plugin<b_peer_plugin> {
    // no override
};

int main()
{
    std::list<std::shared_ptr<base_peer_plugin>> m_extensions;

    a_peer_plugin a;
    b_peer_plugin b;

    m_extensions.push_back(std::make_shared<base_peer_plugin>(a));
    m_extensions.push_back(std::make_shared<base_peer_plugin>(b));
}
```

Reply:
Anyway, I'm just goofing off here. If unused virtual function calls are a concern, then probably it's better to not hook so many functions. Remember I only needed three, and one of them already exists in the code.

Reply:
I don't mind :) C++ can be a little tricky at times, specially under heavy optimization